R/cube-subset.R

```diff
@@ -0,0 +1,235 @@
+#' the categorical variable's categories along one dimension and the MR
+#' dimensions on the other two.
+#'
+#' The complexity is that while the real MR cube includes two dimensions per MR,
+#' we only show the user one dimension per MR which represents only the
+#' `selected` items. This means that every cube has two different
+#' representations, the low dimensional user cube, and the higher dimensional
+#' programmatic cube. This function translates user cube subsets into the higher
+#' dimensional subset. In the case above, the user would see a 2d cube and
+#' subset it with `user_cube[1:2, 1:2]` in order to subset the programmatic cube
+#' we need to translate this to `prog_cube[1:2, 1:2, ]` in order to select the
+#' right variables of the high dimensional cube.
+#' @param x  a Crunch Cube
+#' @param subset a list of array extent indices (for the user-cube)
+#' @param drop whether to drop unnecessary dimensions.
+#' @keywords internal
+#' @return a list of array extent indices (for the real-cube)
+translateCubeIndex <- function(x, subset, drop) {
+    is_selected <- is.selectedDimension(x@dims)
+    #the real cube dimensions
+    prog_names <- names(is_selected)
+    # the user cube dimensions
+    user_names <- prog_names[!is_selected]
+    if (length(prog_names) == length(user_names)) {
+        #no MR variables so no need to translate the subset
+        return(subset)
+    }
+    # This is the main work of the subset translation, just taking the user
+    # supplied subset and projecting them up to real cube dimension
+    out <- as.list(rep(TRUE, length(prog_names)))
+    out[match(user_names, prog_names)] <- subset
+
+    # Dropping MR variables is a bit special. Whenever the user drops the MR
+    # dimension the MR selection dimension is also dropped. This checks if it's
+    # an MR dimension which is being dropped, and assigns an appropriate index
+    # value.
+    if (drop) {
+        for (i in seq_along(prog_names)) {
+            if (i == 1) {
+                next
+            }
+            if (is_selected[i] && #check if MR selection dimension
+                length(out[[i - 1]]) == 1 && # MR response variable is a single number
+                !isTRUE(out[[i - 1]])) {
+                if (x@useNA == "no") {
+                    # This is used by skipMissingCategories below
+                    out[i] <- "mr_select_drop"
+                } else {
+                    # assign index value to "Selected" along the mr_selection dimension
+                    out[i] <- 1
+                }
+            }
+        }
+    }
+    return(out)
+}
+
+
+#' Handle missing categories in CrunchCube
+#'
+#' By default we don't display missing categories. The result is that when the
+#' user subsets a cube with a missing category, we need to translate that subset
+#' to only refer to the non-missing categories. This only occurs if
+#' `cube@useMissing` is set to `"no"`. This function handles this behavior by
```

Comment:
You should add to the sentence that starts "this only occurs if..." that it also could occur if `useNA = "ifAny"` and there aren't any non-0 missing categories.